README.rst
======================
 Ubuntu image builder
======================

This tool is used to build Ubuntu images.  Currently it only builds Snappy
images from a model assertion, but it will be generalized to build more
(eventually all) Ubuntu images.


Requirements
============

Ubuntu 16.04 (Xenial Xerus) is the minimum platform requirement, but Ubuntu
18.04 (Bionic Beaver) or newer is recommended.  Python 3.5 is the minimum Python
version required.  All required third party packages are available in the
Ubuntu archive.

If you want to run the test suite locally, you should install all the build
dependencies named in the `debian/control` file.  The easiest way to do that
is to run::

    $ sudo apt build-dep ./

from the directory containing the `debian` subdirectory.  Alternatively of
course, you can just install the packages named in the `Build-Depends` field.


License
=======

``ubuntu-image`` is licensed under the terms of the GPLv3 and Copyright by
Canonical Ltd.

``ubuntu_image/testing/flake8.py`` and ``ubuntu_image/testing/nose.py`` are
copyright Barry Warsaw and licensed under the terms of the Apache License,
2.0.


Project details
===============

* Project home: https://github.com/CanonicalLtd/ubuntu-image
* Report bugs at: https://bugs.launchpad.net/ubuntu-image
* Git clone: https://github.com/CanonicalLtd/ubuntu-image.git
* Documentation: TBD
* Manual page: man ubuntu-image
  (https://github.com/CanonicalLtd/ubuntu-image/blob/master/ubuntu-image.rst)

The ``gadget.yaml`` specification has moved to `the snapcraft forum`_.

.. _`the snapcraft forum`: https://forum.snapcraft.io/t/the-gadget-snap


Developing
==========

You'll need the `tox <https://pypi.python.org/pypi/tox>`__ tool to run the
test suite (see above for testing requirements).  You can run the full test
suite, including coverage and code quality tests via::

    $ tox

You can run individual tests like this::

    $ tox -e py37-nocov -- -P <pattern>

where *<pattern>* is a Python regular expression matching a test name, e.g.::

    $ tox -e py37-nocov -- -P test_smoke

Pull requests run the same test suite that archive promotion (i.e. -proposed
to release pocket) runs.  You can reproduce this locally by building the
source package (with ``gbp buildpackage -S``) and running::

    $ autopkgtest ubuntu-image_1.9+20.04ubuntu1.dsc -- schroot focal-amd64

with changes to the version number and Ubuntu distroseries as appropriate.


debian/changelog
ubuntu-image (1.11+21.04ubuntu3) UNRELEASED; urgency=medium

  [ Łukasz 'sil2100' Zemczak ]
  * For ubuntu-core images, delegate image customization such as disabling
    console-conf and cloud-init user-data to snap prepare-image.  This way
    those two options should finally start working on UC20 as well.
    (LP: #1875431)
  * debian/control: bump the snapd dependency to a version that will most
    likely have support for the prepare-image customization.

  [ Samuele Pedroni ]
  * Add --factory-image command option.  (LP: #1929236)
  * Add --validation command option.

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Fri, 21 May 2021 18:10:41 +0200

ubuntu-image (1.11+21.04ubuntu2) hirsute; urgency=medium

  * Refreshed manpage.
  * debian/control: remove the explicit grub-common dependency in
    python3-ubuntu-image.  This was added as a workaround for snapd using
    grub-editenv without an explicit dependency - but since long they switched
    to using a native implementation instead.

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Fri, 09 Apr 2021 11:14:38 +0200

ubuntu-image (1.11+21.04ubuntu1) hirsute; urgency=medium

  [ Michael Vogt ]
  * Use "resolved-content" directory provided by newer versions
    of "snap prepare-image". (LP: #1636927)

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Thu, 25 Mar 2021 14:55:14 +0100

ubuntu-image (1.10+20.10ubuntu4) hirsute; urgency=medium

  * Add python3.9 to the test envlists.
  * Minor README fixups and additions, minor test-related compatibility
    changes.

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Wed, 13 Jan 2021 11:27:53 +0100

ubuntu-image (1.10+20.10ubuntu3) hirsute; urgency=medium

  * setup.py:
    - get the version in a way which is compatible with the current
      python-debian

 -- Sebastien Bacher <seb128@ubuntu.com>  Thu, 26 Nov 2020 15:30:53 +0100

ubuntu-image (1.10+20.10ubuntu2) groovy; urgency=medium

  [ Brian Murray ]
  * Fix a typo in README.rst.

  [ Łukasz 'sil2100' Zemczak ]
  * Add a workaround for mkfs.ext4 forcefully sparsifying our classic image
    swapfile (LP: #1900718)

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Tue, 20 Oct 2020 19:40:57 +0200

ubuntu-image (1.10+20.10ubuntu1) groovy; urgency=medium

  [ Łukasz 'sil2100' Zemczak ]
  * Add a quick option to disable console-conf.  (LP: #1866002)
  * Disable some of the unsupported features for UC20 models.  (LP: #1875431)
    - --disable-console-conf
    - --cloud-init
    - Running the post-populate-rootfs hook.
  * Do not create an empty /boot on the system-seed partition.
  * Switch the approach of building UC20 images to ensure that all partitions
    defined in the gadget are considered in image size calculations.
    (LP: #1859973)

  [ Michael Hudson-Doyle ]
  * Add dependency on fdisk now that it is no longer Essential.  (LP: #1877485)

  [ Maciej Borzecki ]
  * Do not overwrite the files that snap prepare-image has created.
    (LP: #1884247)

  [ Dimitri John Ledkov ]
  * Pacify flake8 in groovy.

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Thu, 03 Sep 2020 10:20:28 +0200

ubuntu-image (1.9+20.04ubuntu3) groovy; urgency=medium

  * Newest sfdisk includes a new field of 'sectorsize' in its output, which our
    tests did not expect to see - work-around it by ignoring it for the
    relevant tests.  Fixes FTBFS.

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Fri, 26 Jun 2020 10:17:36 +0200

ubuntu-image (1.9+20.04ubuntu2) groovy; urgency=medium

  [ Michael Hudson-Doyle ]
  * Add dependency on fdisk now that it is no longer Essential. (LP: #1877485)

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Thu, 18 Jun 2020 09:33:47 +0200

ubuntu-image (1.9+20.04ubuntu1) focal; urgency=medium

  [ Łukasz 'sil2100' Zemczak ]
  * Add support for populating a .disk/info file on the target image rootfs.
    (LP: #1856684)
  * Design change: no longer remove /boot/grub from the rootfs for classic
    image builds.  This allows developers to have more freedom on deciding how
    boot configuration should be handled.  (LP: #1817050)
  * Fix the autopkgtest story on focal.  Only for this series, remove the py37
    tox environment as it does not work due to pyparted being 3.8-only.

  [ Maciej Borzecki ]
  * ubuntu_image: update schema validator to allow gadget update specific
    keys & little cleanups.  (LP: #1856903)

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Wed, 29 Jan 2020 16:33:36 +0100

ubuntu-image (1.8+20.04ubuntu2) focal; urgency=medium

  * Fix manifest generation for uc20 builds.  (LP: #1856205)

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Thu, 12 Dec 2019 17:05:59 +0100

ubuntu-image (1.8+20.04ubuntu1) focal; urgency=medium

  [ Łukasz 'sil2100' Zemczak ]
  * Add support for uc20 images and the system-seed partition role.
    (LP: #1847231)

  [ Alfonso Sanchez-Beato ]
  * Add support for LK (little kernel) bootloader (LP: #1845964)

 -- Łukasz 'sil2100' Zemczak <lukasz.zemczak@ubuntu.com>  Tue, 08 Oct 2019 12:56:43 +0200

ubuntu-image (1.7+19.10ubuntu6) eoan; urgency=medium

  * Require isolation-machine for snapcraft specific autopkgtests

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Thu, 12 Sep 2019 12:59:48 +0000

ubuntu-image (1.7+19.10ubuntu5) eoan; urgency=medium

  * Fix previous upload, to also fix E501 long line error

 -- Gianfranco Costamagna <locutusofborg@debian.org>  Tue, 30 Jul 2019 12:24:12 +0200

ubuntu-image (1.7+19.10ubuntu4) eoan; urgency=medium

  * Try to fix F723 syntax error in type comment

 -- Gianfranco Costamagna <locutusofborg@debian.org>  Tue, 30 Jul 2019 11:39:33 +0200

ubuntu-image (1.7+19.10ubuntu3) eoan; urgency=medium

  [ Dimitri John Ledkov ]
  * Use modern apt commands.

 -- Gianfranco Costamagna <locutusofborg@debian.org>  Tue, 30 Jul 2019 08:26:32 +0200

ubuntu-image (1.7+19.10ubuntu2) eoan; urgency=medium

  * Fix E117 over-indended line


release.py
#!/usr/bin/python3

import os
import re
import sys

from contextlib import ExitStack, contextmanager
from debian.changelog import Changelog
from git import Repo
from git.exc import GitCommandError
from subprocess import run
from tempfile import NamedTemporaryFile


@contextmanager
def chdir(path):
    here = os.getcwd()
    try:
        os.chdir(path)
        yield
    finally:
        os.chdir(here)


@contextmanager
def atomic(dst, encoding='utf-8'):
    """Open a temporary file for writing using the given encoding.

    The context manager returns an open file object, into which you can write
    text or bytes depending on the encoding it was opened with.  Upon exit,
    the temporary file is moved atomically to the destination.  If an
    exception occurs, the temporary file is removed.

    :param dst: The path name of the target file.
    :param encoding: The encoding to use for the open file.  If None, then
        file is opened in binary mode.
    """
    directory = os.path.dirname(dst)
    mode = 'wb' if encoding is None else 'wt'
    with ExitStack() as resources:
        fp = resources.enter_context(NamedTemporaryFile(
            mode=mode, encoding=encoding, dir=directory, delete=False))
        yield fp
        os.rename(fp.name, dst)


def update_changelog(repo, series, version):
    # Update d/changelog.
    with ExitStack() as resources:
        debian_changelog = os.path.join(
            repo.working_dir, 'debian', 'changelog')
        infp = resources.enter_context(
            open(debian_changelog, 'r', encoding='utf-8'))
        outfp = resources.enter_context(atomic(debian_changelog))
        changelog = Changelog(infp)
        changelog.distributions = series
        series_version = {
            'hirsute': '21.04',
            'groovy': '20.10',
            'focal': '20.04',
            'bionic': '18.04',
            'xenial': '16.04',
            }[series]
        new_version = '{}+{}ubuntu1'.format(version, series_version)
        changelog.version = new_version
        changelog.write_to_open_file(outfp)
    return new_version


def sru_tracking_bug(repo, sru):
    with ExitStack() as resources:
        debian_changelog = os.path.join(
            repo.working_dir, 'debian', 'changelog')
        infp = resources.enter_context(
            open(debian_changelog, 'r', encoding='utf-8'))
        outfp = resources.enter_context(atomic(debian_changelog))
        changelog = Changelog(infp)
        changelog.add_change('  * SRU tracking number LP: #{}'.format(sru))
        changelog.write_to_open_file(outfp)


def continue_abort(msg='Pausing'):
    print(msg)
    while True:
        answer = input('[c]ontinue, [a]bort? ')
        if answer == 'a':
            print('Aborting!  Fix things manually')
            sys.exit(1)
        elif answer == 'c':
            break


def tag_or_skip(repo, version):
    force = False
    while True:
        answer = input('[t]ag, [f]orce, or [s]kip? ')
        if answer == 's':
            return
        if answer in 'tf':
            if answer == 'f':
                force = True
            break
    repo.create_tag(version, force=force)


def munge_lp_bug_numbers(repo):
    debian_changelog = os.path.join(repo.working_dir, 'debian', 'changelog')
    with ExitStack() as resources:
        infp = resources.enter_context(
            open(debian_changelog, 'r', encoding='utf-8'))
        outfp = resources.enter_context(atomic(debian_changelog))
        changelog = Changelog(infp)
        # Iterate through every line in the top changelog block.  Because we
        # want to modify the existing LP bug numbers, and because the API
        # doesn't give us direct access to those lines, we need to pop the
        # hood, reach in, and manipulate them ourselves.
        for i, line in enumerate(changelog[0]._changes):
            munged = re.sub('LP: #([0-9]+)', 'LP:\\1', line)
            changelog[0]._changes[i] = munged
        changelog.write_to_open_file(outfp)


def make_source_package(working_dir):
    with chdir(working_dir):
        run(['gbp', 'buildpackage', '-S', '-us', '-uc', '--git-ignore-branch'])


def main():
    try:
        working_dir = sys.argv[1]
    except IndexError:
        working_dir = os.getcwd()
    repo = Repo(working_dir)
    assert not repo.bare
    # Start by modifying the master branch.
    print('Updating master...')
    repo.heads.master.checkout()
    # The version number.
    version = input('version: ')
    sru = input('SRU tracking bug: ')
    # Modify the snapcraft.yaml.
    snapcraft_yaml = os.path.join(working_dir, 'snapcraft.yaml')
    with ExitStack() as resources:
        infp = resources.enter_context(
            open(snapcraft_yaml, 'r', encoding='utf-8'))
        outfp = resources.enter_context(atomic(snapcraft_yaml))
        for line in infp:
            if line.startswith('version: '):
                print('version:', '{}+snap1'.format(version), file=outfp)
            else:
                outfp.write(line)
    new_version = update_changelog(repo, 'hirsute', version)
    continue_abort('Pausing for manual review and commit')
    tag_or_skip(repo, new_version)
    make_source_package(working_dir)
    # Now do the Groovy branch.
    repo.git.checkout('groovy')
    # This will almost certainly cause merge conflicts.
    try:
        repo.git.merge('master', '--no-commit')
    except GitCommandError:
        continue_abort('Resolve merge master->groovy conflicts manually')
    munge_lp_bug_numbers(repo)
    sru_tracking_bug(repo, sru)
    new_version = update_changelog(repo, 'groovy', version)
    continue_abort('Pausing for manual review and commit')
    tag_or_skip(repo, new_version)
    make_source_package(working_dir)
    # Now do the Focal branch.
    repo.git.checkout('focal')
    # This will almost certainly cause merge conflicts.
    try:
        repo.git.merge('master', '--no-commit')
    except GitCommandError:
        continue_abort('Resolve merge master->focal conflicts manually')
    munge_lp_bug_numbers(repo)
    sru_tracking_bug(repo, sru)
    new_version = update_changelog(repo, 'focal', version)
    continue_abort('Pausing for manual review and commit')
    tag_or_skip(repo, new_version)
    make_source_package(working_dir)
    # Now do the Bionic branch.
    repo.git.checkout('bionic')
    # This will almost certainly cause merge conflicts.
    try:
        repo.git.merge('master', '--no-commit')
    except GitCommandError:
        continue_abort('Resolve merge master->bionic conflicts manually')
    munge_lp_bug_numbers(repo)
    sru_tracking_bug(repo, sru)
    new_version = update_changelog(repo, 'bionic', version)
    continue_abort('Pausing for manual review and commit')
    tag_or_skip(repo, new_version)
    make_source_package(working_dir)
    # Now do the Xenial branch.
    repo.git.checkout('xenial')
    # This will almost certainly cause merge conflicts.
    try:
        repo.git.merge('master', '--no-commit')
    except GitCommandError:
        continue_abort('Resolve merge master->xenial conflicts manually')
    munge_lp_bug_numbers(repo)
    sru_tracking_bug(repo, sru)
    new_version = update_changelog(repo, 'xenial', version)
    continue_abort('Pausing for manual review and commit')
    tag_or_skip(repo, new_version)
    make_source_package(working_dir)
    # Back to master and create the snap.
    repo.heads.master.checkout()


if __name__ == '__main__':
    main()
